frame/plasm-lockdrop/src/lib.rs

```diff
@@ -0,0 +1,662 @@
+                    }
+                };
+                let claim = Claim { params, amount, .. Default::default() };
+                <Claims>::insert(claim_id, claim);
+            }
+
+            <Requests>::mutate(|requests| requests.push(claim_id));
+            Self::deposit_event(RawEvent::ClaimRequest(claim_id));
+        }
+
+        /// Claim tokens according to lockdrop procedure.
+        #[weight = SimpleDispatchInfo::FixedNormal(1_000_000)]
+        fn claim(
+            origin,
+            claim_id: ClaimId,
+        ) {
+            let _ = ensure_signed(origin)?;
+            ensure!(T::Time::now() < T::LockdropEnd::get(), "lockdrop should be finished");
+
+            let claim = <Claims>::get(claim_id);
+            ensure!(!claim.complete, "claim should be already paid"); 
+
+            let positive = claim.approve.saturating_sub(claim.decline) > T::PositiveVotes::get();
+            let threshold = claim.approve + claim.decline > T::VoteThreshold::get();
+            if !threshold || !positive {
+                Err("this request don't get enough authority confirmations")?
+            }
+
+            // Deposit lockdrop tokens on locking public key.
+            let account = match claim.params {
+                Lockdrop::Bitcoin { public, .. } => T::Account::from(public).into_account()
+            };
+            let amount: BalanceOf<T> = T::BalanceConvert::from(claim.amount).into();
+            T::Currency::deposit_creating(&account, amount);
+
+            // Finalize claim request
+            <Claims>::mutate(claim_id, |claim| claim.complete = true);
+            Self::deposit_event(RawEvent::ClaimComplete(claim_id, account, amount));
+        }
+
+        /// Vote for claim request according to check results. (for authorities only) 
+        #[weight = SimpleDispatchInfo::FixedOperational(10_000)]
+        fn vote(
+            origin,
+            claim_id: ClaimId,
+            approve: bool,
+        ) {
+            let sender = ensure_signed(origin)?;
+            ensure!(Self::is_authority(&sender), "this method for lockdrop authorities only");
+            ensure!(<Claims>::exists(claim_id), "request with this id doesn't exist");
+
+            <Claims>::mutate(&claim_id, |claim|
+                if approve { claim.approve += 1 }
+                else { claim.decline += 1 }
+            );
+
+            Self::deposit_event(RawEvent::ClaimResponse(claim_id, sender, approve));
+        }
+
+        /// Dollar rate oracle method. (for authorities only) 
+        #[weight = SimpleDispatchInfo::FixedOperational(10_000)]
+        fn set_dollar_rate(
+            origin,
+            time: T::Moment,
+            rate: T::DollarRate,
+        ) {
+            let sender = ensure_signed(origin)?;
+            ensure!(Self::is_authority(&sender), "this method for lockdrop authorities only");
+            <DollarRateF<T>>::insert(sender, (time, rate));
```

Comment:
`DollarRateF` will increase without remove.
Then, `DollarRateF<T>>::enuamerate()` is more and heavier as time passes.
I think DollaRate calculating at `on_finalize` or `on_initialize` methods. What's about you?

Reply:
I can erase expired fields during filter enumeration, I guess it's better than touch it on each block.